ivy.el

```diff
@@ -1752,10 +1752,10 @@ See also `ivy-sort-max-size'."
                  (function :tag "Custom function")
                  (repeat (function :tag "Custom function")))))
 
-(defun ivy--sort-function (collection)
+(defun ivy--sort-function (collection &optional ignore-last-state)
   "Retrieve sort function for COLLECTION from `ivy-sort-functions-alist'."
   (let ((entry (cdr (or (assq collection ivy-sort-functions-alist)
-                        (assq (ivy-state-caller ivy-last) ivy-sort-functions-alist)
+                        (when (not ignore-last-state) (assq (ivy-state-caller ivy-last) ivy-sort-functions-alist))
```

Comment:
This looks pretty messy. Why not just bind `ivy-last` around the call to `ivy--sort-function` instead?

BTW, `when`+`not` = `unless`.

Reply:
Because we don't know what `ivy-last` will be, because it's modified by `ivy--reset-state` (at least that's my understanding of it). The only other alternative I see is "inlining" the function into `ivy--reset-state` like this 
```elisp
(sort (or (ivy-state-sort state) (assq caller ivy-sort-functions-alist) (assq t ivy-sort-functions-alist)))
```

but then, we'd lose the sanity checking with `functionp` thet `ivy--sort-function` gives us.

Then again, I'm not an expert on this code, I don't really understand why `ivy--sort-function` accesses the last state if the caller isn't found.

Reply:
> Because we don't know what `ivy-last` will be, because it's modified by `ivy--reset-state`

I'm not sure what you mean. What I mean is the following change to `ivy--reset-state`:

```diff
diff --git a/ivy.el b/ivy.el
index 1b324904..e42b951d 100644
--- a/ivy.el
+++ b/ivy.el
@@ -2226,7 +2226,9 @@ ivy--reset-state
          (dynamic-collection (ivy-state-dynamic-collection state))
          (require-match (ivy-state-require-match state))
          (caller (or (ivy-state-caller state) this-command))
-         (sort (or (ivy-state-sort state) (assoc caller ivy-sort-functions-alist)))
+         (sort (or (ivy-state-sort state)
+                   (let ((ivy-last state))
+                     (ivy--sort-function caller))))
          (initial-input
           (or (ivy-state-initial-input state)
               (let ((init (ivy-alist-setting ivy-initial-inputs-alist caller)))
```

This doesn't make `ivy--sort-function` any more messy than it already is, and AFAICT achieves the same result that you're after. Or am I missing something?

Reply:
> Then again, I'm not an expert on this code, I don't really understand why `ivy--sort-function` accesses the last state if the caller isn't found.

Don't ask me - I try not to think about it. ;)

Reply:
> What I mean is the following change to `ivy--reset-state`:

Or, alternatively:

```diff
diff --git a/ivy.el b/ivy.el
index 1b324904..8ed97898 100644
--- a/ivy.el
+++ b/ivy.el
@@ -2226,7 +2226,9 @@ ivy--reset-state
          (dynamic-collection (ivy-state-dynamic-collection state))
          (require-match (ivy-state-require-match state))
          (caller (or (ivy-state-caller state) this-command))
-         (sort (or (ivy-state-sort state) (assoc caller ivy-sort-functions-alist)))
+         (sort (or (ivy-state-sort state)
+                   (let ((ivy-last state))
+                     (ivy--sort-function collection))))
          (initial-input
           (or (ivy-state-initial-input state)
               (let ((init (ivy-alist-setting ivy-initial-inputs-alist caller)))
```

so that `ivy--sort-function` looks up both `collection` and `caller`.